### ISSUE TITLE
Control over which node runs a job in cluster

### ISSUE THREAD

**Issue by maximiliancsuk** [quartznet/quartznet#1615]
**Is your feature request related to a problem? Please describe.**
I have the following usecase: a cluster setup with two nodes. I'd like to run periodic jobs. The jobs read data from a database, process it, then write data back to the database. To improve performance, I want to keep a cache of intermediate data between job runs. The amount of intermediate data is relatively large, so persisting it to the database between job runs would not make much sense (then I could just not use a cache after all). So want to keep this cached data in-memory.

**A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]***
AFAICT, in a quartz cluster setup, a job runs semi-randomly on one of the nodes and I have no guarantee that a periodically triggered job always runs on the same node. That's why I cannot use in-memory caching between runs as I'd like.

**Describe the solution you'd like**
Ideally, I'd like to be able to specify that a job should be "pinned" to a node: at the first job-run, a node is chosen randomly. But from then on out, consecutive runs of that job always will be executed on the same node as the first run. Only in case of a failure of the node, the job should be switched to the other node.

If that's not possible, I'd also settle for a different approach, where I have to specify the node a job should run on manually.

In general, I'd like to have more control over which node runs a job.

**Describe alternatives you've considered**
I've thought about using an additional local scheduler that is configured to run the job and use that only on one of the nodes. The problem is that this requires manual intervention in case of a node failure to get the jobs running again.

I've also thought about a combination of local and distributed/clustered scheduler instances: a distributed job first picks the "correct" node to run on, then triggers a local job on that node that does the actual work. Problem here is - again - that there is no way to tell quartz to run the local job on a specific node. Variation of that idea: running the distributed job in a high frequency and hope that it gets executed on the correct node, which I'd detect and if so, trigger the local job. Seems less that ideal though...

I might have overlooked something crucial in the docs or misunderstood things... if so, please tell me! :)

**Comment by jafin** [comment]
Have you considered persisting your cache to something like a REDIS store?  It should afford you the speed of maintaining a cache and not be affected by which node runs the Quartz job. ?

I would worry with pinning, that you start to lose the benefits of operating a cluster, i.e. what if the cluster node that is pinned with the cache goes down?  What should occur now?

**Comment by lahma** [comment]
Java version has/had `Quartz Where` which was a paid addon to allow such configuration based on wanted characteristics. Not sure where it went, here's [the announcement](https://www.eweek.com/enterprise-apps/terracotta-enhances-quartz-java-job-scheduler/) though.

**Comment by maximiliancsuk** [comment]
> Have you considered persisting your cache to something like a REDIS store? It should afford you the speed of maintaining a cache and not be affected by which node runs the Quartz job. ?
> 
Yes, I've thought about that, but even a store like REDIS would require marshalling/unmarshalling of the data and moving it between the store and the node all the time. I'd really want the intermediate data to be in-place and in-memory. Plus, I want to keep our setup simple, with the database as the only form of communication between nodes. Introducing another data store would make our setup and testing much more complicated.
 
> I would worry with pinning, that you start to lose the benefits of operating a cluster, i.e. what if the cluster node that is pinned with the cache goes down? What should occur now?
>
If the node that was executing the job goes down, quartz should just "switch" the job to the other node and trigger jobs there from then on out. On the first run on the new node, the cache would be empty, but that's totally ok. The only thing I'd need to think about more is invalidating the cache when it gets stale... but that would be my problem, nothing to do with quartz. :)

**Comment by maximiliancsuk** [comment]
> Java version has/had `Quartz Where` which was a paid addon to allow such configuration based on wanted characteristics. Not sure where it went, here's [the announcement](https://www.eweek.com/enterprise-apps/terracotta-enhances-quartz-java-job-scheduler/) though.

I've seen that in my googling attempts too, but couldn't make any use of it. I assume nothing like that exists in the .Net implementation?

**Comment by lahma** [comment]
> > Java version has/had `Quartz Where` which was a paid addon to allow such configuration based on wanted characteristics. Not sure where it went, here's [the announcement](https://www.eweek.com/enterprise-apps/terracotta-enhances-quartz-java-job-scheduler/) though.
> 
> I've seen that in my googling attempts too, but couldn't make any use of it. I assume nothing like that exists in the .Net implementation?

Nope. It's a non-trivial endeavor to implement it...

**Comment by manguoes** [comment]
+1 for this kind of feature.

I have a use case with multiple nodes/clustering, where an API instance and the Quartz job instance is installed on each node.
When a request is received to the API on node 1, I also want the job to prefer to run on node 1.

**Comment by lahma** [comment]
Being worked on here: https://github.com/quartznet/quartznet/pull/3013 , basically it's preferred node for trigger.

**Comment by lahma** [comment]
Implemented as **node affinity (preferred node)** in #3013, targeting 3.19.0.

Set a trigger's preferred node via `TriggerBuilder.WithPreferredNode(instanceId)` to pin it to a specific cluster node, or `"*"` to auto-pin it to the first node that fires it. It's a strong preference with automatic failover — if the pinned node stops checking in, surviving nodes take over, and auto-pins reset once the owner dies. Backed by `PREFERRED_NODE`/`PREFERRED_NODE_AUTO` columns on `QRTZ_TRIGGERS`; RAMJobStore treats it as metadata only.

Ported to 4.x with docs in #3145, plus clustered integration tests in #3146. See the [Node Affinity tutorial](https://www.quartz-scheduler.net/documentation/quartz-4.x/tutorial/node-affinity.html).
